docs/proposals/dynamic-kubelet-settings.md

```diff
@@ -0,0 +1,129 @@
+
+## Abstract
+
+A proposal for improving Kubelet config by versioning it and making it manageable via the API server.
+
+## Motivation
+
+The Kubelet is currently configured via command-line flags. This is painful for a number of reasons:
+- It makes it difficult to change the way Kubelets are configured in a running cluster, because it is often tedious to change the Kubelet startup configuration.
+- It makes it difficult to manage different Kubelet configurations for different nodes, e.g. if you want to canary a new config.
+- The current lack of versioned Kubelet configuration means that any time we change Kubelet flags, we risk breaking someone's setup.
+- Having the ability to pass config in the same format on-disk as via the API server will make it easier to bootstrap master nodes before the API server is available.
+
+
+## Goals of this design:
+
+- Add a well-defined, versioned configmap object for Kubelet configuration.
+- Add/Remove/Update global and per-node Kubelet config on the fly via the API server.
+- Provide a reliable way to get the configuration of a given Kubelet along with its source (remote or on disk or flags).
+- We want the Kubelet to work with the built-in defaults on all distros we test against.
+- Work towards deprecating flags in favor of on-disk config and config from the API server. Master nodes will be bootstrapped from on-disk config.
+- Make it possible to opt-out of remote configuration e.g. to protect the master from bad config updates.
+
+## Design
+
+### Definitions
+
+- `kubelet-default`: The built-in default config that every Kubelet starts up with.
+- `api-default`: The defaults provided for omitted/nil fields by the API server.
+- `on-disk-config`: `KubeletConfiguration` provided on disk, used as a fallback when the Kubelet cannot cannot contact the API server.
+- `minimal-config`: Conceptually, what the Kubelet needs to start up and contact the API server. This probably consists of the location of the API server and credentials to access the API server, and might be provided via flags or via `on-disk-config`.
+- `global-kubelet-configuration`: the `KubeletConfiguration` configmap set via the API server which is shared by all Kubelets in your cluster except those which have a `<node-name>-kubelet-configuration`.
+- `<node-name>-kubelet-configuration`: the `KubeletConfiguration` configmap specific to the node named `<node-name>`, also set via the API server.
+
+### Well-defined configmap type for Kubelet configuration
+
+- There is an alpha version of the `KubeletConfiguration` type: external in `pkg/componentconfig/v1alpha1/types.go`, internal in `pkg/componentconfig/types.go`.
+- We likely want to group related fields of the `KubeletConfiguration` into substructures. Please note that `v1alpha1` will not be regrouped, but eventually we will start working on this.
+
+@timstclair and I did some brainstorming and came up with some rough potential categories:
+
+---
+- Networking
```

Comment:
We'll toggle entire alpha features with #30003. But each feature probably has some associated knobs to configure, how do we provide that configuration to the node?

With respect to using the `KubeletConfiguration` struct:

I suspect it will be advantageous to avoid field churn in the `KubeletConfiguration` types, especially once componentconfig hits beta, and *especially* once it hits GA. If we e.g. told people to put their alpha knobs in an alpha version of the componentconfig API group, I can imagine questions similar to the following coming up: 
- "my feature is alpha, `KubeletConfiguration` is now in componentconfig v1beta1 (or, heck v1), does that mean I need to create a v2alpha1 for the entire componentconfig API group to work on just my one little feature?"
- "And do I have to do the work to keep that in sync with any changes to the current version of the componentconfig API group (more relevant if something is in beta vs GA)? "
- "And, say pre v1, what happens if someone wants to move the componentconfig API group alpha->beta or beta->GA, and my feature is still alpha and volatile? Would I end up blocking the advancement of the API group? Would I get left behind?"

I think it makes sense to use the versioned API for breaking changes, like doing a major restructuring of a config struct, or deprecating fields. But creating a new version for an API group seems like a much bigger deal than just adding alpha knobs, so, from my perspective, the act of just adding alpha knobs doesn't seem to have an obvious place in the versioning model.

One way to avoid field churn, which has the potential to affect the entire API group, is to provide the configuration data for alpha knobs via annotations on the node, as @derekwaynecarr suggested. 

That still leaves the question of how to transition alpha knobs into the `KubeletConfiguration` structs as they become stable, which affects the API group.

What do you guys think about this? @mikedanese, @bgrant0607 is my understanding of the limitations of the versioned API groups accurate?